### PR TITLE
Feat/withdraw queue

### PR DESCRIPTION
**Pull request by crispymangoes**

### CODE REVIEW

File: test/WithdrawQueue.t.sol [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+        queue.updateWithdrawRequest(cellar, req);
+
+        WithdrawQueue.WithdrawRequest memory savedReq = queue.getUserWithdrawRequest(address(this), cellar);
+
+        assertTrue(savedReq.inSolve == false, "inSolve should be false");
+    }
+
+    // TODO test showing how user originally wants 100 shares withdrawn, then changes it to 50, make sure only 50 can be withdrawn.
+
+    function finishSolve(bytes calldata runData, uint256, uint256 assetApprovalAmount) external {
```
Comment (0xEinCodes) [inline test/WithdrawQueue.t.sol:236]: So is this just checking what happens if the solver attempts to solve only half of the amount?

File: src/modules/withdraw-queue/WithdrawQueue.sol [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+
+        // Setup meta data.
+        metaData = new SolveMetaData[](users.length);
+
+        for (uint256 i; i < users.length; ++i) {
+            WithdrawRequest memory request = userWithdrawRequest[users[i]][share];
+
+            metaData[i].user = users[i];
+
+            if (block.timestamp > request.deadline) {
```
Comment (0xEinCodes) [inline src/modules/withdraw-queue/WithdrawQueue.sol:259]: Let's walk through this together, just want to make sure I understand the flags and solver process.

**Review by 0xEinCodes — COMMENTED** [review]
Looking good man, just want to check over some aspects with you over a 15 minute chat or something.

File: src/modules/withdraw-queue/WithdrawQueue.sol [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+    struct SolveData {
+        ERC20 asset;
+        uint8 assetDecimals;
+        uint8 shareDecimals;
+    }
+
+    // ========================================= CONSTANTS =========================================
+
+    // ========================================= GLOBAL STATE =========================================
+    mapping(address => mapping(ERC4626 => WithdrawRequest)) public userWithdrawRequest;
```
[the remark this answers is not in the document]
Comment (crispymangoes) [reply src/modules/withdraw-queue/WithdrawQueue.sol:72]: Yeahp!

File: src/modules/withdraw-queue/WithdrawQueue.sol [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+                emit RequestFulfilled(
+                    users[i],
+                    address(share),
+                    request.sharesToWithdraw,
+                    assetsToUser,
+                    block.timestamp
+                );
+
+                // Set shares to withdraw to 0.
+                request.sharesToWithdraw = 0;
```
[the remark this answers is not in the document]
Comment (crispymangoes) [reply src/modules/withdraw-queue/WithdrawQueue.sol:229]: The only mutative functions have reentrancy gaurds on them, so moving these values to the top just increases the function complexity cuz now we need to save these values in memory before zeroing them out, then use the memory values in the computations.

File: test/WithdrawQueue.t.sol (hunk 1 above)
Comment (crispymangoes) [reply test/WithdrawQueue.t.sol:751]: So this test had a user submitting a request to withdraw X shares, then they decided to reduce it to be x/2 shares. This test verifies that solvers can only solve x/2 shares.

File: src/modules/withdraw-queue/WithdrawQueue.sol [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+        return true;
+    }
+
+    // TODO there is a front running attack like the ERC20 approval attack
+    // User already has a non zero request, they submit a TX to increase it
+    // Attacker sees it, front runs it, solves it, then lets users TX goes through and solves again.
+    function updateWithdrawRequest(ERC4626 share, WithdrawRequest calldata userRequest) external nonReentrant {
+        WithdrawRequest storage request = userWithdrawRequest[msg.sender][share];
+
+        request.deadline = userRequest.deadline;
```
[the remark this answers is not in the document]
Comment (0xEinCodes) [reply src/modules/withdraw-queue/WithdrawQueue.sol:146]: Fair points, but what if:

1. The UI fails to perform how we want it to, and then users are exposed to bad withdraw requests (with bad pricing for example)? UX-wise, users would be pretty upset. It wouldn't be the responsibility of the protocol or anything but still bad UX no? The alternative is some checks to ensure that it doesn't allow that, which means more gas... maybe something to discuss w/ Joe and Sun?

^^ we discussed this and will further discuss w/ Jordan & Front-End team.

File: src/modules/withdraw-queue/WithdrawQueue.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (0xEinCodes) [reply src/modules/withdraw-queue/WithdrawQueue.sol:230]: We discussed this offline and your point is valid. All good!

File: src/modules/withdraw-queue/WithdrawQueue.sol [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+    //============================== ERRORS ===============================
+
+    error WithdrawQueue__UserRepeated();
+    error WithdrawQueue__RequestDeadlineExceeded();
+    error WithdrawQueue__UserNotInSolve();
+    error WithdrawQueue__NoShares();
+
+    //============================== EVENTS ===============================
+
+    event RequestUpdated(
```
[the remark this answers is not in the document]
Comment (0xEinCodes) [reply src/modules/withdraw-queue/WithdrawQueue.sol:86]: We covered this, and we are good!

**Comment by 0xEinCodes** [comment]
We discussed any lingering aspects. All looks good to me to pass on to the external auditor to begin working with while small TODOs are resolved (and small spelling haha 😊). Good job man!

**Pull request merged**